_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -0,0 +1,215 @@
+
+partof: binary-compatibility
+permalink: /overviews/core/:title.html
+---
+
+**By Jacob Wang**
+
+## Introduction
+
+A diverse and comprehensive set of libraries is important to any productive software ecosystem. While it is easy to develop and distribute Scala libraries, good library authorship goes
+beyond just writing code and publishing it.
+
+In this guide, we will cover the important topic of **Binary Compatibility**:
+
+* How binary incompatibility can cause production failures in your applications
+* How to avoid breaking binary compatibility
+* How to reason about and communicate the impact of their code changes
+
+Before we start, let's understand how code is compiled and executed on the Java Virtual Machine (JVM).
+
+## The JVM execution model
+
+Scala is compiled to a platform-independent format called **JVM bytecode** and stored in `.class` files. These class files are collated in JAR files for distribution.
+
+When some code depends on a library, its compiled bytecode references the library's bytecode. The library's bytecode is referenced by its class/method signatures and loaded lazily
+by the JVM classloader during runtime. If a class or method matching the signature is not found, an exception is thrown. 
+
+As a result of this execution model:
+
+* We need to provide the JARs of every library used in our dependency tree when starting an application since the library's bytecode is only referenced, not merged into its user's bytecode
+* A missing class/method problem may only surface after the application has been running for a while, due to lazy loading.
+
+Common exceptions from classloading failures includes 
+`InvocationTargetException`, `ClassNotFoundException`, `MethodNotFoundException`, and `AbstractMethodError`.
+
+Let's illustrate this with an example:
+
+Consider an application `App` that depends on `A` which itself depends on library `C`. When starting the application we need to provide the class files
+for all of `App`, `A` and `C` (something like `java -cp App.jar:A.jar:C.jar:. MainClass`). If we did not provide `C.jar` or if we provided a `C.jar` that does not contain some classes/methods
+which `A` calls, we will get classloading exceptions when our code attempts to invoke the missing classes/methods.
+
+These are what we call **Binary Incompatibility Errors** -- errors that happen when the compiled bytecode references a name that cannot be resolved during runtime.
+
+Before we look at how to avoid binary incompatibility errors, let us first
+establish some key terminologies we will be using for the rest of the guide.
+
+## What are Evictions, Source Compatibility and Binary Compatibility?
+
+### Evictions
+When a class is needed during execution, the JVM classloader loads the first matching class file from the classpath (any other matching class files are ignored).
+Because of this, having multiple versions of the same library in the classpath is generally undesirable:
```

Comment:
This mixes up different things. Consider replacing that paragraph with something more obviously accurate, such as

> Unless using special tools such as OSGI), only one version of any given library can be loaded in the same JVM. If several are provided on the classpath, the contents of the first one will shadow the later ones.

Reply:
What's inaccurate about the paragraph? I want to minimize the amount of keywords mentioned especially when they doesn't help the key message of the guide (e.g. OSGI)

Reply:
@jatcwang I think Sébastien's suggestion here makes sense, just make sure you add a paragraph for "(such as OSGi)" so that it has less importance. But the wording is superior in precision :smile:.